fdbserver/DataDistribution.actor.cpp

```diff
@@ -3519,7 +3577,77 @@ ACTOR Future<Void> monitorStorageServerRecruitment(DDTeamCollection* self) {
 	}
 }
 
+ACTOR Future<Void> checkAndRemoveInvalidLocalityAddr(DDTeamCollection* self) {
+	state int checkCount = 0;
+	state Transaction tr(self->cx);
+
+	loop {
+		try {
+			wait(delay(SERVER_KNOBS->DD_CHECK_INVALID_LOCALITY_DELAY));
```

Comment:
Is there a way we could short-circuit the delay in the event of a policy change? In particular, if someone misconfigures their cluster and wants to restore things quickly, one course of action they can take is to undo the configuration change. Ideally, the cluster would react quickly to that and return to correct operation.

One possibility is to just clear out the set whenever the policy changes and let things work themselves out. Maybe a little better is to terminate the wait and check the localities immediately when the policy changes.

Reply:
Another possible optimization is to immediately restore (or check) a process whenever it's detected in something like `waitServerListChange` with a new process ID.

Reply:
Once configuration is changed, such as the replication policy changes, the `moveKeysLockOwnerKey` will be changed, which will shutdown the old DD and start a new DD. This will clear up the stale invalidLocality set.

Reply:
> Another possible optimization is to immediately restore (or check) a process whenever it's detected in something like `waitServerListChange` with a new process ID.

The problem is that if the processID is not changed when part of locality info (say data_hall) is changed, this won't include the process either. 
If we check if a process's locality is valid in `waitServerListChange `, this will cause unnecessary performance overhead for a cluster with correct locality configuration.

Reply:
That wasn't meant to catch every case, but just to more quickly and cheaply catch a straightforward case

Reply:
It's meant as an optimization, though, and it probably covers a rather unlikely case, so it may not be worth it.

Reply:
I thought about the optimization and have a new doubt:
1) The `waitServerListChange` does not monitor process ID; instead, it monitors serverID and check if a new serverID is added;
2) In order to remove the invalidLocality address when a new serverID is added, a server has to be added to the invalidLocality address first, which is prevented by DD.
So there is no easy way to solve this.

One way is to add `waitProcessListChange` actor to monitor the change of processes / workers and to add a `versionKey` watch to the processList. Whenever the process list (ie.., worker list) changes, we re-check the invalidLocality set. 

This needs change beyond DD. I feel it may not be a priority for this PR.
My opinion of priorities is:
1) Fixing DD's misconfigured locality issue;
2) Adding the simulation test for the misconfigured locality issue;
3) Fixing tLog's misconfigured locality issue;
4) Reducing the latency of recovery from misconfigured locality issue.